compiler/rustc_query_system/src/query/caches.rs

```diff
@@ -100,16 +103,173 @@ where
     }
 }
 
-pub struct VecCache<K: Idx, V> {
-    cache: Lock<IndexVec<K, Option<(V, DepNodeIndex)>>>,
+struct Slot<V> {
+    // We never construct &Slot<V> so it's fine for this to not be in an UnsafeCell.
+    value: V,
+    // This is both an index and a once-lock.
+    //
+    // 0: not yet initialized.
+    // 1: lock held, initializing.
+    // 2..u32::MAX - 2: initialized.
+    index_and_lock: AtomicU32,
 }
 
 impl<K: Idx, V> Default for VecCache<K, V> {
     fn default() -> Self {
-        VecCache { cache: Default::default() }
+        VecCache {
+            buckets: Default::default(),
+            key: PhantomData,
+            len: Default::default(),
+            present: Default::default(),
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+struct SlotIndex {
+    bucket_idx: usize,
+    entries: usize,
+    index_in_bucket: usize,
+}
+
+impl SlotIndex {
+    #[inline]
+    fn from_index(idx: u32) -> Self {
```

Comment:
Have you tried to write a branchless version of this?

Reply:
The result of this on x86 at least is branchless (with conditional moves): https://rust.godbolt.org/z/x9s3aexYh